py/dml/ctree.py

```diff
@@ -1270,12 +1270,19 @@ def make(cls, site, lh, rh):
                 rh_indices = [idx.value for idx in rh.indices]
                 return mkBoolConstant(site, (lh.node is rh.node
                                              and lh_indices == rh_indices))
+            if isinstance(lh, NullExpr) or isinstance(rh, NullExpr):
+                if isinstance(lh, NullExpr) and isinstance(rh, NullExpr):
+                    return mkBoolConstant(site, True)
+                elif (isinstance(lhtype, (TPtr, TArray))
+                      and isinstance(rhtype, (TPtr, TArray))):
+                    return mkBoolConstant(site, False)
```

Comment:
this is incorrect? NULL can be equal to non-constant pointers.

Please add unit test to ctree_test!

Reply:
String literals are constant. Method references converted to function pointers are constant. That's it.
Any PR which allows a pointer to be considered constant by DMLC _and_ can be NULL should be rejected wholesale.

Reply:
Not obvious, consider `cast(0, foo_t*) == NULL`. Perhaps the `NullExpr` class should document that it's the only constant null-valued pointer expression?

Reply:
If we add code to allow `cast(0, foo_t*)` to be constant we'd just add a case to `mkCast` that would create a `NullExpr`.
Now, a more interesting question would be `cast(1, foo_t *) - 1`!
For one, do we want non-zero pointer literals to be constant? My answer is "maybe, whatever."
For another, would we want pointer offsets of constant pointer be offset? My answer would be "certainly not offsets of pointer literals. Maybe for strings if we decide to do interning for string literals."

Reply:
We should make a thing constant if it's technically feasible to do so and there is a real use case for it. Other things can also be considered constant if that leads to a simpler definition of constantness.

Reply:
Now added documentation to `NullConstant` saying that expression class must be used for any null-valued pointer-typed expression considered constant by DMLC.

Reply:
Still no test in ctree_test.py?